numpy/core/src/multiarray/descriptor.c

```diff
@@ -3754,69 +3754,91 @@ descr_repeat(PyObject *self, Py_ssize_t length)
 }
 
 static PyObject *
-descr_subscript(PyArray_Descr *self, PyObject *op)
+_check_has_fields(PyArray_Descr *self)
 {
-    PyObject *retval;
-
-    if (!PyDataType_HASFIELDS(self)) {
-        PyObject *astr = arraydescr_str(self);
+    if (PyDataType_HASFIELDS(self)) return (PyObject *) self;
+    PyObject *astr = arraydescr_str(self);
 #if defined(NPY_PY3K)
-        PyObject *bstr = PyUnicode_AsUnicodeEscapeString(astr);
-        Py_DECREF(astr);
-        astr = bstr;
+    PyObject *bstr = PyUnicode_AsUnicodeEscapeString(astr);
+    Py_DECREF(astr);
+    astr = bstr;
 #endif
-        PyErr_Format(PyExc_KeyError,
+    PyErr_Format(PyExc_KeyError,
                 "There are no fields in dtype %s.", PyBytes_AsString(astr));
-        Py_DECREF(astr);
+    Py_DECREF(astr);
+    return NULL;
+}
+
+static PyObject *
+_subscript_by_name(PyArray_Descr *self, PyObject *op)
+{
+    PyObject *obj = PyDict_GetItem(self->fields, op);
+    PyObject *descr;
+    PyObject *s;
+
+    if (obj == NULL) {
+        if (PyUnicode_Check(op)) {
+            s = PyUnicode_AsUnicodeEscapeString(op);
+        }
+        else {
+            s = op;
+        }
+
+        PyErr_Format(PyExc_KeyError,
+                "Field named \'%s\' not found.", PyBytes_AsString(s));
+        if (s != op) {
+            Py_DECREF(s);
+        }
+        return NULL;
+    }
+    descr = PyTuple_GET_ITEM(obj, 0);
+    Py_INCREF(descr);
+    return descr;
+}
+
+static PyObject *
+_subscript_by_index(PyArray_Descr *self, Py_ssize_t i)
+{
+    PyObject *name = PySequence_GetItem(self->names, i);
+    if(name == NULL) {
+        PyErr_Format(PyExc_IndexError,
+                     "Field index %d out of range.", i);
```

Comment:
Am I right in thinking that there's no need to clear the error before overwriting it?